fmpz_poly/div_series_basecase.c

```diff
@@ -31,59 +43,147 @@ _fmpz_poly_div_series_basecase(fmpz * Q, const fmpz * A, slong Alen,
                 _fmpz_vec_set(Q, A, Alen);
             else
                 _fmpz_vec_neg(Q, A, Alen);
-        } else
+        }
+        else
         {
-            fmpz_init(r);
-
+            slong i;
             for (i = 0; i < Alen; i++)
-            {   
-                fmpz_fdiv_qr(Q + i, r, A + i, B + 0);
-
-                if (!fmpz_is_zero(r))
-                {
-                    fmpz_clear(r);
-
-                    flint_printf("Not an exact division\n");
-                    flint_abort();
-                }
-            }
-
-            fmpz_clear(r);
+                fmpz_divexact_checked(Q + i, A + i, B);
         }
 
         _fmpz_vec_zero(Q + Alen, n - Alen);
     }
+    else if (Alen == 1 && fmpz_is_pm1(B))
+    {
+        _fmpz_poly_inv_series_basecase(Q, B, Blen, n);
+        if (!fmpz_is_one(A))
+            _fmpz_vec_scalar_mul_fmpz(Q, Q, n, A);
+    }
     else
     {
-        slong i, j;
+        slong i, j, nsmall;
+        char *Bbits;
+        slong b, bits, Qbits;
+        TMP_INIT;
+
+        TMP_START;
 
         if (fmpz_is_pm1(B))
         {
             if (fmpz_is_one(B))
                 fmpz_set(Q, A);
             else
                 fmpz_neg(Q, A);
-        } else
+        }
+        else
         {
-            fmpz_init(r);
+            fmpz_divexact_checked(Q, A, B);
+        }
 
-            fmpz_fdiv_qr(Q + 0, r, A + 0, B + 0);
+        /* Bbits[i] = max(bits(B[0]), ..., bits(B[i])), as long as coeffs are small */
+        Bbits = TMP_ALLOC(Blen);
+        Bbits[0] = fmpz_bits(B);
 
-            if (!fmpz_is_zero(r))
-            {
-                fmpz_clear(r);
+        /* Maximum bits of all Q coefficients encountered so far */
+        Qbits = fmpz_bits(Q);
 
-                flint_printf("Not an exact division\n");
-                flint_abort();
-            }
+        /* We have small coefficients for i < nsmall */
+        for (nsmall = 0; nsmall < Blen; nsmall++)
+        {
+            b = B[nsmall];
+
+            if (COEFF_IS_MPZ(b))
+                break;
+
+            b = FLINT_ABS(b);
+            if (nsmall == 0 || (b >> Bbits[nsmall - 1]))
+                Bbits[nsmall] = FLINT_BIT_COUNT(b);
+            else
+                Bbits[nsmall] = Bbits[nsmall - 1];
         }
 
         for (i = 1; i < n; i++)
         {
-            fmpz_mul(Q + i, B + 1, Q + i - 1);
+            if (i >= nsmall || Qbits > FLINT_BITS - 2 || Bbits[i] > FLINT_BITS - 2)
+            {
+                /* Can't use fast code. */
+                bits = WORD_MAX;
+            }
+            else
+            {
+                /* Can maybe use fast code; bound bits. */
+                b = FLINT_MIN(i, Blen - 1);
```

Comment:
Is B[0] used below. Could you just use i - 1 here?

Reply:
I don't follow. min(i, Blen - 1) is the exact number of terms being added.

Reply:
I might have gotten confused. I thought the loop would go through the first i terms of B,  but not B[0]. That would be i - 1 terms. I guess I am off by one somewhere?

Reply:
My mistake. I see the j < i + 1 below.